### PR TITLE
Fix policy migration issues

### LINKED ISSUE #7799: Investigate and fix relationship issue during the rbac migration
|Wazuh branch|
|---|
| dev-rbac-upgrade-and-migration |

Hello team,

It seems the rbac migration development branch does not work as expected when an existing custom resource (such as a policy) created by a user matches with one of the default resources that must be created during the database upgrade process and that custom resource is linked to other custom resources (such as a custom role).

A similar issue was addressed in https://github.com/wazuh/wazuh/pull/7594. It seems the changes applied in that PR conflicts with some of the changes applied to the rbac migration development branch.

## Scenario
Having an existing policy in the user range (id > 99) that matches with a default Wazuh policy during the `rbac` database update process and the user has created a relationship between the policy and one or more custom roles.

## Expected behaviour
The existing policy should be replaced by the default policy but the relationship between that role and the policy should be kept (updating its policy id if necessary)

## Current behaviour
Its seems the link between the policy and the role is being removed after the upgrade process.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### PR DESCRIPTION
**Pull request by CarlosRS9**
Hello team,

This PR closes #7799. This PR fixes the policy and roles-policies migration issue introduced by #7594 as these changes were not compatible with the current `orm` implementation. 

During the migration process, if a user-created Policy conflicts with a default Policy (because they share the same policy body but have different names) that user Policy won't be inserted in the new database and any Role related to that user Policy will have their relationship updated to point to the default Policy instead.

The manual testing performed is described in #7799.

## Changes introduced:
- The inconsistencies between the changes recently introduced in master branch and the changes in the new `orm` module have been fixed.
- The changes introduced in #7594 to the `insert_data_from_yaml` function have been reverted as they are no longer needed with the current implementation of the `orm` module.
- `get_data` function now uses a list instead of a set to ensure the results are kept sorted.
- `check_default` parameter has been removed from the migration function as during a migration every resource must be inserted overwriting the reserved id range checks.
- A bug related to `run_as` in the migration function has been fixed.
- `test_add_new_default_policies` and `test_migrate_default_policies` have been removed from `test_security` as they are not intended to work with the new `orm` module.- A new test called `test_database_manager_migrate_data` has been added to `test_orm` instead.

### How the new unit test works
The new `test_database_manager_migrate_data` will try to replicate the issue described in #7799:
1. Create a database with the default resources
2. Add custom user resources to it.
3. Modify some default policies to ensure there will be a conflict between those resources and the ones they will be added later.
4. Creates new relationships between user roles and the modified policies.
5. Initiate the migration process.
6. Check the migrated database to see if the policies were fixed while updating it's relationships.
7. Check the rest of the resources are still present.

## Unit tests results
```
============================================= test session starts ==============================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 1747 items                                                                                           

framework/wazuh/core/cluster/dapi/tests/test_dapi.py .......................                             [  1%]
framework/wazuh/core/cluster/tests/test_cluster.py ...................                                   [  2%]
framework/wazuh/core/cluster/tests/test_common.py .......                                                [  2%]
framework/wazuh/core/cluster/tests/test_control.py sssss                                                 [  3%]
framework/wazuh/core/cluster/tests/test_local_client.py .                                                [  3%]
framework/wazuh/core/cluster/tests/test_utils.py .......                                                 [  3%]
framework/wazuh/core/cluster/tests/test_worker.py ............ss.s                                       [  4%]
framework/wazuh/core/tests/test_active_response.py ..............                                        [  5%]
framework/wazuh/core/tests/test_agent.py ............................................................... [  8%]
...........................................................................................              [ 14%]
framework/wazuh/core/tests/test_cdb_list.py ......................................                       [ 16%]
framework/wazuh/core/tests/test_common.py .......                                                        [ 16%]
framework/wazuh/core/tests/test_configuration.py ....................................                    [ 18%]
framework/wazuh/core/tests/test_decoder.py ................                                              [ 19%]
framework/wazuh/core/tests/test_input_validator.py ...                                                   [ 19%]
framework/wazuh/core/tests/test_logtest.py ..                                                            [ 19%]
framework/wazuh/core/tests/test_manager.py .................                                             [ 20%]
framework/wazuh/core/tests/test_pyDaemonModule.py ......                                                 [ 21%]
framework/wazuh/core/tests/test_results.py ........................................                      [ 23%]
framework/wazuh/core/tests/test_rootcheck.py ..........                                                  [ 24%]
framework/wazuh/core/tests/test_rule.py ......................                                           [ 25%]
framework/wazuh/core/tests/test_stats.py ....                                                            [ 25%]
framework/wazuh/core/tests/test_syscollector.py ...                                                      [ 25%]
framework/wazuh/core/tests/test_utils.py ............................................................... [ 29%]
........................................................................................................ [ 35%]
....................................................                                                     [ 38%]
framework/wazuh/core/tests/test_vulnerability.py .                                                       [ 38%]
framework/wazuh/core/tests/test_wazuh_queue.py ...................                                       [ 39%]
framework/wazuh/core/tests/test_wazuh_socket.py ....................                                     [ 40%]
framework/wazuh/core/tests/test_wdb.py .....................                                             [ 41%]
framework/wazuh/rbac/tests/test_auth_context.py ..                                                       [ 41%]
framework/wazuh/rbac/tests/test_decorators.py .......................................................... [ 45%]
...............................................                                                          [ 47%]
framework/wazuh/rbac/tests/test_default_configuration.py ............................................... [ 50%]
........                                                                                                 [ 51%]
framework/wazuh/rbac/tests/test_orm.py .......................................................           [ 54%]
framework/wazuh/rbac/tests/test_preprocessor.py ...........                                              [ 54%]
framework/wazuh/tests/test_active_response.py ............                                               [ 55%]
framework/wazuh/tests/test_agent.py .................................................................... [ 59%]
............................                                                                             [ 61%]
framework/wazuh/tests/test_cdb_list.py .....................................................             [ 64%]
framework/wazuh/tests/test_ciscat.py .................................                                   [ 65%]
framework/wazuh/tests/test_cluster.py .....ss                                                            [ 66%]
framework/wazuh/tests/test_decoder.py ...................................                                [ 68%]
framework/wazuh/tests/test_group.py ........                                                             [ 68%]
framework/wazuh/tests/test_logtest.py ......                                                             [ 69%]
framework/wazuh/tests/test_manager.py .................................                                  [ 71%]
framework/wazuh/tests/test_mitre.py .................................................................... [ 74%]
........................................................................................................ [ 80%]
........                                                                                                 [ 81%]
framework/wazuh/tests/test_rootcheck.py .................................................                [ 84%]
framework/wazuh/tests/test_rule.py ........................................................              [ 87%]
framework/wazuh/tests/test_sca.py .......                                                                [ 87%]
framework/wazuh/tests/test_security.py ................................................................. [ 91%]
...........................................                                                              [ 93%]
framework/wazuh/tests/test_stats.py ................                                                     [ 94%]
framework/wazuh/tests/test_syscheck.py ........................                                          [ 96%]
framework/wazuh/tests/test_syscollector.py ............                                                  [ 96%]
framework/wazuh/tests/test_task.py ............................                                          [ 98%]
framework/wazuh/tests/test_vulnerability.py ..........................                                   [100%]

============================ 1737 passed, 10 skipped, 26 warnings in 417.34 seconds ============================
```

```
============================================= test session starts ==============================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/api
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 225 items                                                                                            

api/api/test/test_alogging.py ..........                                                                 [  4%]
api/api/test/test_authentication.py ..........                                                           [  8%]
api/api/test/test_configuration.py .........................................                             [ 27%]
api/api/test/test_util.py ...................................                                            [ 42%]
api/api/test/test_validator.py ......................................................................... [ 75%]
........................................................                                                 [100%]
=================================== 225 passed, 16 warnings in 1.32 seconds ====================================
```

## API integration tests

```
test_agent_DELETE_endpoints.tavern.yaml 
	 6 passed, 1 warnings

test_agent_GET_endpoints.tavern.yaml 
	 91 passed, 1 warnings

test_agent_POST_endpoints.tavern.yaml 
	 5 passed, 1 warnings

test_agent_PUT_endpoints.tavern.yaml 
	 9 passed, 1 warnings

test_rbac_black_agent_endpoints.tavern.yaml 
	 40 passed, 1 warnings 

test_rbac_white_agent_endpoints.tavern.yaml 
	 40 passed, 1 warnings

test_cluster_endpoints.tavern.yaml 
	 44 passed, 1 xpassed, 1 warnings

test_rbac_black_cluster_endpoints.tavern.yaml 
	 17 passed, 1 warnings

test_rbac_white_cluster_endpoints.tavern.yaml 
	 17 passed, 1 warnings

test_manager_endpoints.tavern.yaml 
	 31 passed, 1 warnings

test_rbac_black_manager_endpoints.tavern.yaml 
	 15 passed, 1 warnings

test_rbac_white_manager_endpoints.tavern.yaml 
	 15 passed, 1 warnings

test_rbac_black_security_endpoints.tavern.yaml 
	 25 passed, 1 warnings

test_rbac_white_security_endpoints.tavern.yaml 
	 25 passed, 1 warnings

test_security_DELETE_endpoints.tavern.yaml 
	 15 passed, 1 warnings

test_security_GET_endpoints.tavern.yaml 
	 18 passed, 1 warnings

test_security_POST_endpoints.tavern.yaml 
	 7 passed, 1 warnings

test_security_PUT_endpoints.tavern.yaml 
	 9 passed, 1 warnings

```

### CODE REVIEW

File: framework/wazuh/rbac/orm.py [hunk 1]
```diff
@@ -2662,59 +2663,29 @@ def insert_data_from_yaml(self, database: str, resource_type: ResourceType = Res
...
-                                                .index(policy_id)
-
-                                        pm.delete_policy(policy_id=policy_id)
-                                        pm.add_policy(name=policy_name, policy=policy, resource_type=resource_type,
-                                                      check_default=False)
-                                        policy_id = pm.get_policy(policy_name)['id']
-                                        for role, position in new_positions.items():
-                                            rpm.add_role_to_policy(policy_id=policy_id, role_id=role, position=position,
-                                                                   force_admin=True)
+                        pm.add_policy(name=f'{d_policy_name}_{name}', policy=policy, check_default=False)
```
Comment (vicferpoy) [inline framework/wazuh/rbac/orm.py:2687]:
```suggestion
                        pm.add_policy(name=f'{d_policy_name}_{name}', policy=policy, resource_type=resource_type, check_default=False)
```

This was adding default policies as `user`.

**Review by vicferpoy — CHANGES_REQUESTED** [review]
If I link some policies to a new role

```sql
sqlite> select * from roles_policies where role_id=100;
47|100|4|0|2021-03-29 10:32:26.366825
48|100|5|1|2021-03-29 10:32:26.366825
49|100|6|2|2021-03-29 10:32:26.366825
50|100|7|3|2021-03-29 10:32:26.366825

```

And then force a migration
```
2021/03/29 10:37:33 INFO: Checking RBAC database integrity...
2021/03/29 10:37:33 INFO: /var/ossec/api/configuration/security/rbac.db file was detected
2021/03/29 10:37:33 INFO: RBAC database migration required. Current version is 0 but it should be 40400
2021/03/29 10:37:33 INFO: Upgrading RBAC database to 40400 version
2021/03/29 10:37:35 INFO: /var/ossec/api/configuration/security/rbac.db database upgraded successfully.
2021/03/29 10:37:35 INFO: RBAC database integrity check finished successfully.
2021/03/29 10:37:40 INFO: Listening on 0.0.0.0:55000..

```

The role loses all its policies
```
sqlite> select * from roles_policies where role_id=100;
sqlite> 

```

This needs to be fixed as soon as possible as it was working before.

File: framework/wazuh/rbac/orm.py [hunk 2]
```diff
@@ -2747,22 +2717,25 @@ def insert_data_from_yaml(self, database: str, resource_type: ResourceType = Res
...
+                            column >= from_id).order_by(column).all()
                     elif to_id:
-                        partial_result = set(self.sessions[source].query(table).filter(column <= to_id).all())
-                    result = partial_result if result == set() else result.intersection(partial_result)
+                        partial_result = self.sessions[source].query(table).filter(
+                            column <= to_id).order_by(column).all()
+                    else:
+                        partial_result = self.sessions[source].query(table).order_by(column).all()
+                    # We need to keep results sorted so intersection between partial_results must be resolved manually.
+                    result = partial_result if result == list() else [x for x in result if x in partial_result]
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply framework/wazuh/rbac/orm.py]: This function has been reworked

File: framework/wazuh/rbac/orm.py [hunk 3]
```diff
@@ -2797,16 +2769,30 @@ def get_data(table, *args):
...
-                                          policy_id=policy.id,
-                                          check_default=check_default)
+                status = policy_manager.add_policy(name=policy.name,
+                                                   policy=json.loads(policy.policy),
+                                                   created_at=policy.created_at,
+                                                   policy_id=policy.id,
+                                                   check_default=False)
+                # If the user's policy has the same body as an existing default policy it won't be inserted and its
+                # role-policy relationships will be linked to that default policy instead to replace it.
+                if status == SecurityError.ALREADY_EXIST:
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply framework/wazuh/rbac/orm.py]: Added

File: framework/wazuh/rbac/orm.py (hunk 1 above)
Comment (CarlosRS9) [reply framework/wazuh/rbac/orm.py]: Fixed

**Comment by CarlosRS9** [comment]
> If I link some policies to a new role
> 
> ```sql
> sqlite> select * from roles_policies where role_id=100;
> 47|100|4|0|2021-03-29 10:32:26.366825
> 48|100|5|1|2021-03-29 10:32:26.366825
> 49|100|6|2|2021-03-29 10:32:26.366825
> 50|100|7|3|2021-03-29 10:32:26.366825
> ```
> 
> And then force a migration
> 
> ```
> 2021/03/29 10:37:33 INFO: Checking RBAC database integrity...
> 2021/03/29 10:37:33 INFO: /var/ossec/api/configuration/security/rbac.db file was detected
> 2021/03/29 10:37:33 INFO: RBAC database migration required. Current version is 0 but it should be 40400
> 2021/03/29 10:37:33 INFO: Upgrading RBAC database to 40400 version
> 2021/03/29 10:37:35 INFO: /var/ossec/api/configuration/security/rbac.db database upgraded successfully.
> 2021/03/29 10:37:35 INFO: RBAC database integrity check finished successfully.
> 2021/03/29 10:37:40 INFO: Listening on 0.0.0.0:55000..
> ```
> 
> The role loses all its policies
> 
> ```sql
> sqlite> select * from roles_policies where role_id=100;
> sqlite> 
> ```
> 
> This needs to be fixed as soon as possible as it was working before.

This was happening due to a bug related to the resources' relationship migration. This part has been reworked to fix it. 

The current version now will check if a relationship resource, such as `role_policies`, is related to a default resource and if it does, it check if the given default resource is present in the new database even if its ID has changed (by looking for its name). If the default resource is present with a different ID the relationship will be added with the updated ID. If the resource is no longer present the relationship will be omited and won't be added to the new database.

**Example removing a default policy:**

Some `roles_policies` resources were created using a user-created role with id `100`:
```
select * from roles_policies where role_id=100;
49|100|4|0|2021-03-29 15:18:10.962402
50|100|5|1|2021-03-29 15:18:10.962402
51|100|6|2|2021-03-29 15:18:10.962402
```

Then the default policy with id `5` was removed from the default `policies.yaml` and a migration was forced. After the migration, that particular relationship was removed but the rest were kept:
```
select * from roles_policies where role_id=100;
47|100|4|0|2021-03-29 15:28:11.819731
48|100|5|1|2021-03-29 15:28:11.819731
```

**Example changing the id of a default policy:**

Some `roles_policies` resources were created using a user-created role with id `100`:
```
select * from roles_policies where role_id=100;
49|100|4|0|2021-03-29 15:18:10.962402
50|100|5|1|2021-03-29 15:18:10.962402
51|100|6|2|2021-03-29 15:18:10.962402
```

Then the default policies with id `5` and `6` were altered in the default `policies.yaml` to ensure it would have a different id next time and a migration was forced. After the migration, that particular relationship was updated with the new id while the rest were kept:
```
select * from roles_policies where role_id=100;
51|100|4|2|2021-03-29 15:56:52.553351
49|100|34|0|2021-03-29 15:56:52.553351
50|100|35|1|2021-03-29 15:56:52.553351
```

**Review by vicferpoy — APPROVED** [review]
LGTM!

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**